File: src/lib/rubeno/lib/base_it.rb
module Rubeno
  # BaseIt extends BaseAction for Describe-It pattern.
  # Its can mix mutations and assertions, unlike BDD's When which only does mutations.
  class BaseIt < BaseAction
    def initialize(name, it_cb)
      super(name, it_cb)
    end
    
    def perform_action(store, action_cb, test_resource, artifactory = nil)
      # This should be implemented by subclasses
      raise NotImplementedError, "Subclasses must implement perform_action"
    end
    
    # Alias perform_action to performIt
    def perform_it(store, it_cb, test_resource, artifactory = nil)
      perform_action(store, it_cb, test_resource, artifactory)
    end
    
    # Alias for Describe-It pattern
    def it(store, it_cb, test_resource_configuration, artifactory = nil)
      perform_action(store, it_cb, test_resource_configuration, artifactory)
    end
  end
end

File: src/lib/kafe/src/main/java/kafe/BaseIt.java
package kafe;

import java.util.function.Function;

/**
 * BaseIt extends BaseAction for Describe-It pattern.
 * Its can mix mutations and assertions, unlike BDD's When which only does mutations.
 */
public abstract class BaseIt<R> extends BaseAction<Object, R> {
    public BaseIt(String name, Function<R, R> itCb) {
        super(name, itCb);
    }
    
    // Alias performAction to itDoes for Describe-It pattern
    @Override
    public abstract R performAction(
        R store,
        Function<R, R> actionCB,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory
    );
    
    // Alias test to itTest for Describe-It pattern
    public R itTest(R store, ITTestResourceConfiguration testResourceConfiguration) throws Exception {
        return super.test(store, testResourceConfiguration);
    }
}
